unity/src/main/java/io/delta/unity/UCCatalogManagedCommitter.java

```diff
@@ -0,0 +1,196 @@
+      throws CommitFailedException {
+    final String stagedCommitFilePath = commitMetadata.getNewStagedCommitFilePath();
+
+    return OperationTimer.timeOperation(
+        logger,
+        "Write staged commit file",
+        ucTableId,
+        () -> {
+          try {
+            // Do not use Put-If-Absent for staged commit files since we assume that UUID-based
+            // commit files are globally unique, and so we will never have concurrent writers
+            // attempting to write the same commit file.
+
+            engine
+                .getJsonHandler()
+                .writeJsonFileAtomically(
+                    stagedCommitFilePath, finalizedActions, true /* overwrite */);
+
+            // TODO: [delta-io/delta#5021] Use FileSystemClient::getFileStatus API instead
+            return FileStatus.of(stagedCommitFilePath);
+          } catch (IOException ex) {
+            throw new CommitFailedException(
+                true /* retryable */,
+                false /* conflict */,
+                "Failed to write staged commit file due to: " + ex.getMessage(),
+                ex);
+          }
+        });
+  }
+
+  private void commitToUC(CommitMetadata commitMetadata, FileStatus kernelFileStatus)
+      throws CommitFailedException {
+    OperationTimer.timeOperation(
+        logger,
+        "Commit to UC",
+        ucTableId,
+        () -> {
+          final boolean isDisown =
+              commitMetadata.getCommitType()
+                  == CommitMetadata.CommitType.CATALOG_DOWNGRADE_TO_FILESYSTEM;
+
+          try {
+            ucClient.commit(
+                ucTableId,
+                tablePath.toUri(),
+                Optional.of(getCommitStructPayload(commitMetadata, kernelFileStatus)),
+                Optional.empty() /* lastKnownBackfilledVersion */,
+                isDisown,
+                Optional.empty() /* newMetadata */,
+                Optional.empty());
+            return null;
+          } catch (io.delta.storage.commit.CommitFailedException cfe) {
+            throw storageCFEtoKernelCFE(cfe);
+          } catch (IOException ex) {
+            throw new CommitFailedException(
+                true /* retryable */, false /* conflict */, ex.getMessage(), ex);
+          } catch (UCCommitCoordinatorException ucce) {
+            throw new CommitFailedException(
+                false /* retryable */, false /* conflict */, ucce.getMessage(), ucce);
+          }
+        });
+  }
+
+  private Commit getCommitStructPayload(
+      CommitMetadata commitMetadata, FileStatus kernelFileStatus) {
+    return new Commit(
+        commitMetadata.getVersion(),
+        kernelFileStatusToHadoopFileStatus(kernelFileStatus),
+        // commitMetadata validates the ICT is present if writing to a catalogManaged table
+        commitMetadata.getCommitInfo().getInCommitTimestamp().get());
+  }
+
+  private static org.apache.hadoop.fs.FileStatus kernelFileStatusToHadoopFileStatus(
```

Comment:
side note, I didn't realize we were baking Hadoop dependencies in Kernel commits?